File: src/main/java/com/example/oauth/service/AuthorizationTokenService.java
package com.example.oauth.service;

import com.example.oauth.config.token.AuthorizationTokenProperties;
import com.example.oauth.model.TokenAttributes;
import com.example.oauth.model.TokenType;
import com.example.oauth.repository.SocialUserRepository;
import com.example.oauth.repository.TokenStoreRepository;
import com.example.oauth.repository.model.SocialUser;
import com.example.oauth.repository.model.TokenStore;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import javax.mail.AuthenticationFailedException;
import java.util.Date;

@Service
@RequiredArgsConstructor
public class AuthorizationTokenService {
    private static final String JWT_SUBJECT_PREFIX = "USER/";

    private final AuthorizationTokenProperties tokenProperties;
    private final TokenStoreRepository tokenStoreRepository;
    private final SocialUserRepository userRepository;

    public String createAccessToken(SocialUser socialUser) {
        return createToken(TokenType.ACCESS_TOKEN, socialUser);
    }

    public String createRefreshToken(SocialUser socialUser, String accessToken) {
        String refreshToken = createToken(TokenType.REFRESH_TOKEN, socialUser);
        Long existTokenStoreId = tokenStoreRepository.findByUserId(socialUser.getId())
                .orElseGet(TokenStore::new)
                .getId();

        TokenStore tokenStore = TokenStore.builder()
                .id(existTokenStoreId)
                .userId(socialUser.getId())
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .createTime(System.currentTimeMillis())
                .build();
        tokenStoreRepository.save(tokenStore);
        return refreshToken;
    }

    private String createToken(TokenType tokenType, SocialUser socialUser) {
        Date now = new Date();
        Date expiration = new Date(now.getTime() + tokenProperties.getTokenExpirationMsec(tokenType));

        TokenAttributes tokenAttributes = TokenAttributes.extract(tokenType, socialUser);
        assert tokenAttributes != null;

        return Jwts.builder()
                .setSubject(JWT_SUBJECT_PREFIX + socialUser.getId())
                .setIssuedAt(now)
                .setExpiration(expiration)
                .setClaims(tokenAttributes.getAttributesMap())
                .signWith(SignatureAlgorithm.HS256, tokenProperties.getTokenSecret(tokenType))
                .compact();
    }

    public SocialUser getPseudoSocialUserFromToken(TokenType tokenType, String token) throws AuthenticationFailedException {
        try {
            Claims claims = Jwts.parser()
                    .setSigningKey(tokenProperties.getTokenSecret(tokenType))
                    .parseClaimsJws(token)
                    .getBody();

            TokenAttributes tokenAttributes = TokenAttributes.restore(tokenType, claims);
            assert tokenAttributes != null;
            return tokenAttributes.getPseudoSocialUser();
        } catch (Exception e) {
            throw new AuthenticationFailedException();
        }
    }

    public boolean validateToken(TokenType tokenType, String token) {
        try {
            Jwts.parser()
                    .setSigningKey(tokenProperties.getTokenSecret(tokenType))
                    .parseClaimsJws(token);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public String refreshAccessToken(String expiredAccessToken, String refreshToken) throws AuthenticationFailedException {
        if (refreshToken == null) {
            throw new AuthenticationFailedException();
        }
        SocialUser pseudoSocialUser = getPseudoSocialUserFromToken(TokenType.REFRESH_TOKEN, refreshToken);
        TokenStore tokenStore = tokenStoreRepository.findByUserId(pseudoSocialUser.getId())
                .orElseThrow(AuthenticationFailedException::new);

        if (validateRefreshToken(tokenStore, expiredAccessToken, refreshToken)) {
            SocialUser socialUser = userRepository.findById(pseudoSocialUser.getId())
                    .orElseThrow(AuthenticationFailedException::new);
            String accessToken = createToken(TokenType.ACCESS_TOKEN, socialUser);
            reflectNewAccessTokenToTokenStore(tokenStore, accessToken);
            return accessToken;
        } else {
            throw new AuthenticationFailedException();
        }
    }

    private boolean validateRefreshToken(TokenStore tokenStore, String expiredAccessToken, String refreshToken) {
        return expiredAccessToken.equals(tokenStore.getAccessToken())
                && refreshToken.equals(tokenStore.getRefreshToken());
    }

    private void reflectNewAccessTokenToTokenStore(TokenStore tokenStore, String accessToken) {
        tokenStore.setAccessToken(accessToken);
        tokenStore.setUpdateTime(System.currentTimeMillis());
        tokenStoreRepository.save(tokenStore);
    }
}

File: src/main/java/com/example/oauth/service/AuthorizationTokenService.kt
package com.example.oauth.service

import com.example.oauth.config.properties.AuthorizationTokenProperties
import com.example.oauth.model.token.TokenAttributes
import com.example.oauth.model.token.TokenType
import com.example.oauth.repository.SocialUserRepository
import com.example.oauth.repository.TokenStoreRepository
import com.example.oauth.repository.model.SocialUser
import com.example.oauth.repository.model.TokenStore
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import org.springframework.stereotype.Service
import java.util.*
import javax.mail.AuthenticationFailedException

@Service
class AuthorizationTokenService(
    private val tokenProperties: AuthorizationTokenProperties,
    private val tokenStoreRepository: TokenStoreRepository,
    private val userRepository: SocialUserRepository,
) {

    companion object {
        private const val JWT_SUBJECT_PREFIX = "USER/"
    }

    fun createAccessToken(socialUser: SocialUser): String {
        return createToken(TokenType.ACCESS_TOKEN, socialUser)
    }

    fun createRefreshToken(socialUser: SocialUser, accessToken: String): String {
        val refreshToken = createToken(TokenType.REFRESH_TOKEN, socialUser)
        val existTokenStoreId = tokenStoreRepository.findByUserId(socialUser.id!!)?.id
        val tokenStore = TokenStore(
            id = existTokenStoreId,
            userId = socialUser.id!!,
            accessToken = accessToken,
            refreshToken = refreshToken,
            createTime = System.currentTimeMillis(),
        )

        tokenStoreRepository.save(tokenStore)
        return refreshToken
    }

    private fun createToken(tokenType: TokenType, socialUser: SocialUser): String {
        val now = Date()
        val expiration = Date(now.time + tokenProperties.tokenExpirationMsec(tokenType))
        val tokenAttributes = TokenAttributes.extract(tokenType, socialUser)
        return Jwts.builder()
            .setSubject(JWT_SUBJECT_PREFIX + socialUser.id)
            .setIssuedAt(now)
            .setExpiration(expiration)
            .setClaims(tokenAttributes.attributesMap)
            .signWith(SignatureAlgorithm.HS256, tokenProperties.tokenSecret(tokenType))
            .compact()
    }

    @Throws(AuthenticationFailedException::class)
    fun getPseudoSocialUserFromToken(tokenType: TokenType, token: String): SocialUser {
        return try {
            val claims = Jwts.parser()
                .setSigningKey(tokenProperties.tokenSecret(tokenType))
                .parseClaimsJws(token)
                .body

            val tokenAttributes = TokenAttributes.restore(tokenType, claims)
            tokenAttributes.pseudoSocialUser
        } catch (e: Exception) {
            throw AuthenticationFailedException(e.message)
        }
    }

    fun validateToken(tokenType: TokenType, token: String): Boolean {
        return try {
            Jwts.parser()
                .setSigningKey(tokenProperties.tokenSecret(tokenType))
                .parseClaimsJws(token)

            true
        } catch (e: Exception) {
            false
        }
    }

    @Throws(AuthenticationFailedException::class)
    fun refreshAccessToken(expiredAccessToken: String, refreshToken: String): String {
        val pseudoSocialUser = getPseudoSocialUserFromToken(TokenType.REFRESH_TOKEN, refreshToken)
        val tokenStore = tokenStoreRepository.findByUserId(pseudoSocialUser.id!!)

        return if (validateRefreshToken(tokenStore!!, expiredAccessToken, refreshToken)) {
            val socialUser = userRepository.findById(pseudoSocialUser.id!!)
                .orElseThrow { AuthenticationFailedException("Fail to find social user : ${pseudoSocialUser.id}") }
            val accessToken = createToken(TokenType.ACCESS_TOKEN, socialUser)
            reflectNewAccessTokenToTokenStore(tokenStore, accessToken)

            accessToken
        } else {
            throw AuthenticationFailedException("Fail to validate refresh token : $tokenStore")
        }
    }

    private fun validateRefreshToken(
        tokenStore: TokenStore, expiredAccessToken: String, refreshToken: String
    ): Boolean {
        return expiredAccessToken == tokenStore.accessToken && refreshToken == tokenStore.refreshToken
    }

    private fun reflectNewAccessTokenToTokenStore(tokenStore: TokenStore, accessToken: String) {
        tokenStore.let {
            it.accessToken = accessToken
            it.updateTime = System.currentTimeMillis()
            tokenStoreRepository.save(it)
        }
    }

}